test_conformance/non_uniform_work_group/TestNonUniformWorkGroup.cpp

```diff
@@ -243,75 +243,134 @@ static const char *KERNEL_FUNCTION =
   NL "}"
   NL ;
 
-TestNonUniformWorkGroup::TestNonUniformWorkGroup (const cl_device_id &device, const cl_context &context,
-  const cl_command_queue &queue, const cl_uint dims, const size_t *globalSize, const size_t *localSize, const size_t *buffersSize,
-  const size_t *globalWorkOffset, const size_t *reqdWorkGroupSize)
-  : _device(device), _context(context), _queue(queue), _dims (dims) {
-
-  if (globalSize == NULL || dims < 1 || dims > 3) {
-    //throw std::invalid_argument("globalSize is NULL value.");
-    // This is method of informing that parameters are wrong.
-    // It would be checked by prepareDevice() function.
-    // This is used because of lack of exception support.
-    _globalSize[0] = 0;
-    return;
-  }
+TestNonUniformWorkGroup::TestNonUniformWorkGroup(
+    const cl_device_id &device, const cl_context &context,
+    const cl_command_queue &queue, const cl_uint dims, size_t *globalSize,
+    const size_t *localSize, const size_t *buffersSize,
+    const size_t *globalWorkOffset, const size_t *reqdWorkGroupSize)
+    : _device(device), _context(context), _queue(queue), _dims(dims)
+{
 
-  cl_uint i;
-  _globalWorkOffset_IsNull = true;
-  _localSize_IsNull = true;
+    if (globalSize == NULL || dims < 1 || dims > 3)
+    {
+        // throw std::invalid_argument("globalSize is NULL value.");
+        // This is method of informing that parameters are wrong.
+        // It would be checked by prepareDevice() function.
+        // This is used because of lack of exception support.
+        _globalSize[0] = 0;
+        return;
+    }
 
-  setGlobalWorkgroupSize(globalSize);
-  setLocalWorkgroupSize(globalSize,localSize);
-  for (i = _dims; i < MAX_DIMS; i++) {
-    _globalSize[i] = 1;
-  }
+    // For OpenCL-3.0 support for non-uniform workgroups is optional, it's still
+    // useful to run these tests since we can verify the behavior of the
+    // get_enqueued_local_size() builtin for uniform workgroups, so we round up
+    // the global size to insure uniform workgroups on those 3.0 devices.
+    // We only need to do this when localSize is non-null, otherwise the driver
+    // will select a value for localSize which will be uniform on devices that
+    // don't support non-uniform work-groups.
+    if (nullptr != localSize && get_device_cl_version(device) == Version(3, 0))
```

Comment:
I think this should be `>= Version(3, 0)` to ensure that future device versions remain compatible here.

Even if we were to make non-uniform a requirement in future versions of the spec, the `CL_DEVICE_NON_UNIFORM_WORK_GROUP_SUPPORT` query would still need to be supported and would therefore always return `CL_TRUE` in that case.